### DISCUSSION TITLE
How to set a custom attribute for each spline point?

### DISCUSSION THREAD

**Discussion by robhawkes** [JacquesLucke/animation_nodes#1801] [Q&A]
Hello,

I've recently started digging into Animation Nodes and combining them with the latest Geometry Nodes features from Blender 3.0. Specifically, I'm trying to store custom attributes for each point on a spline that has been created in Animation Nodes, which I can then get for each point from within Geometry Nodes, and hopefully also within material shaders.

For some background, I'm pulling data dynamically from a JSON file using a Script Node so I can't do this in Geometry Nodes solely. And even if I could, there are aspects of Animation Nodes which I would still like to use.

I've as yet been unable to find a way to store custom attributes for each point on a spline. The only attributes that seem to be possible to set per point are the radius, tilt and handles. In an ideal world I'd be able to set custom attributes on each spline point, such as a timestamp, etc.

For now I've been able to hack around this by storing the timestamp floats in the radius attribute, as I wasn't using that attribute.

![image](https://user-images.githubusercontent.com/22612/139536295-a1e9fce8-8e8b-4997-877b-e7eb83310ff0.png)
![image](https://user-images.githubusercontent.com/22612/139536540-3dbe604a-97a0-4914-9458-692e927c9e25.png)

This shows up correctly in the geometry spreadsheet and the radius (timestamp) values are also accessible within Geometry Nodes, and also transfer to the points spreadsheet when a Curve to Points node is applied.

![image](https://user-images.githubusercontent.com/22612/139536458-03408876-4e50-4c3f-b8e9-302a2fbd7c99.png)

Is there a method that I'm missing that allows for storing arbitrary attributes on spline points? It seems that Geometry Nodes supports this as of Blender 3.0 so it'd be great to be able to also set them via Animation Nodes too.

Thanks,
Robin

**Comment by OmarEmaraDev** [comment]
Unfortunately, it seems Blender either doesn't support or doesn't expose custom attributes for splines. Only meshes, point clouds, and hairs are supported. It is weird because support doesn't seem to be there in the kernel methods either, so I am tempted to say this isn't supported. But I will look into it.

**Reply by robhawkes** [reply]
If it helps, this is what I read [in the latest release notes](https://wiki.blender.org/wiki/Reference/Release_Notes/3.0/Nodes_Physics#Curve_Nodes) that made me believe it's possible in Geometry Nodes now:

![image](https://user-images.githubusercontent.com/22612/139539963-2bb0a6d2-5bae-4814-8aa9-0d365421ae7b.png)


Blender feature ticket: https://developer.blender.org/rB627f3571271e

**Reply by OmarEmaraDev** [reply]
@robhawkes Judging by the code in the differential you linked, it seems custom attributes are runtime structures. So they could be used inside GN but not written to disk or interfaced with python. So we would have to add custom attributes to curves first.

**Reply by OmarEmaraDev** [reply]
While at it, since you are testing daily builds, are Animation Nodes node colors brighter and blown out than usual or is it just me?

**Reply by Gerstmann-Bradley** [reply]
> While at it, since you are testing daily builds, are Animation Nodes node colors brighter and blown out than usual or is it just me?

Recently I think something happens with blender's general theme or related stuff. so this phenomenon of blown out color occurs everywhere.

**Reply by OmarEmaraDev** [reply]
@Gerstmann-Bradley So we should probably adjust our defaults to take that into account. Maybe it is a gamma correction change.

**Reply by Gerstmann-Bradley** [reply]
> @Gerstmann-Bradley So we should probably adjust our defaults to take that into account. Maybe it is a gamma correction change.

yes it will be definitely great. I am not sure what they did exactly, but I think it's relevant to this ticket: https://developer.blender.org/rBbfec984cf82a6

**Reply by robhawkes** [reply]
> @robhawkes Judging by the code in the differential you linked, it seems custom attributes are runtime structures. So they could be used inside GN but not written to disk or interfaced with python. So we would have to add custom attributes to curves first.

Thanks for looking into this. At least I know this isn't currently supported so I can continue working around it for the time being.

**Reply by 3DSinghVFX** [reply]
> While at it, since you are testing daily builds, are Animation Nodes node colors brighter and blown out than usual or is it just me?

Yes, it is also true for me. They set the color value for GN nodes `0.188235`.
